pkg/types/provider_median.go

```diff
@@ -17,7 +17,7 @@ type MedianProvider interface {
 
 type PluginMedian interface {
 	// NewMedianFactory returns a new ReportingPluginFactory. If provider implements GRPCClientConn, it can be forwarded efficiently via proxy.
-	NewMedianFactory(ctx context.Context, provider MedianProvider, dataSource, juelsPerFeeCoin median.DataSource, errorLog ErrorLog) (ReportingPluginFactory, error)
+	NewMedianFactory(ctx context.Context, provider MedianProvider, dataSource, juelsPerFeeCoin, gasPrice median.DataSource, errorLog ErrorLog) (ReportingPluginFactory, error)
```

Comment:
What is gas price used for? That seems like a chain-specific detail that shouldn't leak out here.

Reply:
[Example of it on Starknet](https://github.com/smartcontractkit/chainlink-starknet/blob/develop/contracts/src/ocr2/aggregator.cairo#L234)

Reply:
Not all chains require it (like EVM chains), so if it is nil then ocr will not include it in a transmission

Reply:
Oh, is the point that the actual gas price value for each round has to be agreed on?
Is `Gas` general enough? Or is there a more chain-agnostic term for execution units?

Reply:
>Not all chains require it (like EVM chains), so if it is nil then ocr will not include it in a transmission

🤔 I wonder if we're operating at the wrong level of abstraction. This suggests that whether or not to include it is chain-specific, so any behaviour related to it should happen in the relayer.

Wdyt about always calling the datasource, but only including it in the report if it's solana? That would allow us to use the ReportEncoder (which is part of the relayer) as the abstraction point.

Reply:
(I might be jumping the gun btw -- I don't see any code here which actually instantiates a gasPrice datasource and passes it into the median plugin.)

Reply:
Yeah, exactly. the gas price is another value that has to be agreed upon in OCR (so it's similar to JuelesPerCoinFee in that regard). 

I've heard the use of the word `fee` being used (during the txm generalization) when referring to the total gas cost for a transaction. But, `gasPrice` in this case means the price of a single unit of gas so "fee" wouldn't quite work

Reply:
> > Not all chains require it (like EVM chains), so if it is nil then ocr will not include it in a transmission
> 
> 🤔 I wonder if we're operating at the wrong level of abstraction. This suggests that whether or not to include it is chain-specific, so any behaviour related to it should happen in the relayer.
> 
> Wdyt about always calling the datasource, but only including it in the report if it's solana? That would allow us to use the ReportEncoder (which is part of the relayer) as the abstraction point.

It'd be unnecessary to include a data source that isn't used. That means we'd require node ops to include a data source that did nothing for all evm ocr2 pipelines. 

The issue is precisely that whether or not to include a data source of `gasPrice` depends on the chain.

Yeah, there is no code here that instantiates the gasPrice datasource because I haven't gotten to that yet 😅 . Just wanted to get something working first

Reply:
Do other chain families all use the word `gas` though? I can't actually find any other terminology, so maybe nbd :shrug:

Reply:
cosmos and ethereum L2s do. I believe solana operates differently and calls them "tx fees"

Reply:
>It'd be unnecessary to include a data source that isn't used. That means we'd require node ops to include a data source that did nothing for all evm ocr2 pipelines.

>The issue is precisely that whether or not to include a data source of gasPrice depends on the chain.

Can you explain how the gas price data source is going to be instantiated? We are actively working to put all chain-specific details in the relayer, and I want to make sure that we aren't taking a step back here.

Reply:
The data source is another pipeline so it would be instantiated here https://github.com/smartcontractkit/chainlink/pull/10519/files#diff-7358f3249bf57875121f3bcb6211722cf2bee2f9b411dfff89a1974c48c5b67eR109. 

A chain either has one or it doesn't so it's either nil or it's not. There's https://github.com/smartcontractkit/chainlink-relay/pull/166/files#diff-28ea9d6d17fb7499fbbd0f83828a8335ac28eacff629256d75b261bdb2ee123aR59 also when the loopp plugin serves the underlying data source. Would this be a step backward? 

```
This suggests that whether or not to include it is chain-specific, so any behaviour related to it should happen in the relayer.
```

^ If data sources were to be instantiated in the relayer, that means configuration would also occur in the relayer. The infra doesn't exist, but if it were to exist you'd pass in pipeline configs from core to the relay and relay would instantiate them.

Reply:
> Why is a pipeline being used to retrieve this data?

We reimburse `transmit` calls on chain. In order to do that correctly a reimbursement has to be calculated. This is easy in EVM where we can access `tx.gasPrice` in Solidity as well as measure `gasleft()` at the start and end of execution. But in basically every non-EVM chain this data isn't sourceable and we need to do two things:
- come up with reasonable constants to approximate gas use based on lots and lots of transactions
- include a medianized gasPrice in the report (we match the `juelsPerFeeCoin` source usage)

We were able to avoid this second one because:
- on Terra the gas prices were fixed and rarely changed based on some external API price list
- on Solana the TX cost was fixed (for now, but this changes with fee markets)

This `gasPrice` source is needed for both Starknet and Cosmos.